### PR TITLE
Make versioned types for webhook admission config

### PR DESCRIPTION
**Pull request by deads2k**
Versioned webhook admission config type as promised in https://github.com/kubernetes/kubernetes/pull/54414.  

@kubernetes/sig-api-machinery-pr-reviews 
@ericchiang as promised.  fyi.

```yaml
kind: AdmissionConfiguration
apiVersion: apiserver.k8s.io/v1alpha1
plugins:
- name: GenericAdmissionWebhook
  configuration:
    kind: WebhookAdmission
    apiVersion: apiserver.config.k8s.io/v1alpha1
    kubeConfigFile: /path/to/my/file
```

`ADMISSION_CONTROL_CONFIG_FILE=../foo.yaml hack/local-up-cluster.sh`

### CODE REVIEW

**Comment by lavalamp** [comment]
/lgtm
/approve

How does one actually pass in the config file in a flag?

Is it better to have a config file that tells you where to find the kubeconfig file, or a config file that just embeds the kubeconfig file? yaml doesn't make embedding yaml too awkward...

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/kubeconfig.go [hunk 1]
```diff
@@ -18,27 +18,48 @@ package config
 
 import (
 	"io"
+	"io/ioutil"
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/kubeconfig.go]: no empty line

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1alpha1/register.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+package v1alpha1
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// GroupName is the group name use in this package
+const GroupName = "webhook.admission.apiserver.k8s.io"
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1alpha1/register.go]: Against our group name inflation: `config.admission.apiserver.k8s.io`.

**Comment by deads2k** [comment]
> Is it better to have a config file that tells you where to find the kubeconfig file, or a config file that just embeds the kubeconfig file? yaml doesn't make embedding yaml too awkward...

It's a decision to make about which files contain sensitive information and which ones don't.  Since you pass a single file in to configure all of the admission chain, I don't think that file should itself be sensitive since that has support repercussions.  By referencing a separate file, you're able to have the config be safe.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1alpha1/register.go (hunk 2 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1alpha1/register.go]: > Against our group name inflation: config.admission.apiserver.k8s.io.

I'd be ok with that if there were a way to add to a scheme.  Then `admission.apiserver.k8s.io` with specific types like `WebhookConfiguration` makes sense.  I guess I'm building up my own scheme, so I could choose my favorite types and include those.

@liggitt preference?  The common case would still work (that's what I'm doing) and the advanced case could be made to work by a really clever person.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1alpha1/types.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+limitations under the License.
+*/
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Configuration provides configuration for the ResourceQuota admission controller.
```
Comment (jennybuckley) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1alpha1/types.go]: ResourceQuota?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/types.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+limitations under the License.
+*/
+
+package webhookadmission
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Configuration provides configuration for the ResourceQuota admission controller.
```
Comment (jennybuckley) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/types.go]: ResourceQuota?

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/types.go]: I guess we know where this was copied from :)

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1alpha1/register.go (hunk 2 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1alpha1/register.go]: There was a big discussion about this, I thought the judgement was xxx.config.k8s.io, under the assumption that you can look at ".config.k8s.io" and *know* that it's not a "real" api object served by the system.

**Comment by lavalamp** [comment]
> It's a decision to make about which files contain sensitive information and which ones don't. Since you pass a single file in to configure all of the admission chain, I don't think that file should itself be sensitive since that has support repercussions. By referencing a separate file, you're able to have the config be safe.

We already have 2 layers though? We have the overall admission options file, which gives the file path of the generic admission webhook config path (btw will we duplicate that for mutating web hooks or use the same one? Using the same one would be way simpler), which gives the file path of the kubeconfig file.

It seems like we could have the needed separation with just 2 layers? I'm just trying to think of what happens when people start trying to use secrets & configmaps to self host apiserver.

...I guess three layers isn't that bad if you're using a config map because the first two can go in the same config map.

**Comment by liggitt** [comment]
> We have the overall admission options file, which gives the file path of the generic admission webhook config path

You can inline the specific admission plugin's config blob into the admission options file instead of referencing a separate file

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go

That lets you have just two levels if you want

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1alpha1/register.go (hunk 2 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1alpha1/register.go]: > There was a big discussion about this, I thought the judgement was xxx.config.k8s.io, under the assumption that you can look at ".config.k8s.io" and know that it's not a "real" api object served by the system.

Ok.  The existing apiserver config file group doesn't match (it predated the others), but I'm fine adjusting this here.

**Comment by deads2k** [comment]
Updated group name, added example file to description, made the scheme more easily configurable.

**Comment by deads2k** [comment]
@jennybuckley see description for example admission config file.

**Comment by jennybuckley** [comment]
Given that GenericAdmissionWebhook is being split and [renamed](https://github.com/kubernetes/kubernetes/pull/54892#discussion_r151510312) what will this line of the config look like?
```
...
- name: GenericAdmissionWebhook 
...
```

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/types.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// WebhookAdmission provides configuration for the webhook admission controller.
+type WebhookAdmission struct {
+	metav1.TypeMeta
+
+	// KubeConfigFile is the path to the kubeconfig file.
+	KubeConfigFile string
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/types.go]: if this is relative, what it is resolved relative to?

**Comment by caesarxuchao** [comment]
@jennybuckley  I sent a [PR](https://github.com/kubernetes/kubernetes/pull/55988/commits/3a64dfb079f0da7b38d4efe543c5c72bd596eeed) to rename the `GenericAdmissionWebhook` to `ValidatingAdmissionWebhook`. I think we will set entries for both `ValidatingAdmissionWebhook` and `MutatingAdmissionWebhook` in the config file.

**Review by caesarxuchao — COMMENTED** [review]
Two questions and otherwise lgtm.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/kubeconfig.go [hunk 6]
```diff
@@ -18,27 +18,48 @@ package config
...
+	"k8s.io/apimachinery/pkg/apimachinery/registered"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission"
+	"k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/install"
+)
+
+var (
+	groupFactoryRegistry = make(announced.APIGroupFactoryRegistry)
+	registry             = registered.NewOrDie("")
```
[the remark this answers is not in the document]
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/kubeconfig.go]: > Are scheme and codecs enough? Do you need the registry and groupFactoryRegistry?

No.  We just need to be able to decode them.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/register.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var (
+	SchemeBuilder = runtime.NewSchemeBuilder(addKnownTypes)
+	AddToScheme   = SchemeBuilder.AddToScheme
+)
+
+// GroupName is the group name use in this package
+const GroupName = "apiserver.config.k8s.io"
```
[the remark this answers is not in the document]
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/register.go]: > I think it should be webhook.admission.apiserver.k8s.io, just like you put in the comment // +groupName=webhook.admission.apiserver.k8s.io. The name should be plugin-specific.

The type is specific, but the group is consistent with our discussion here: https://github.com/kubernetes/kubernetes/pull/55812#discussion_r151370163

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1alpha1/doc.go [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +k8s:deepcopy-gen=package
+// +k8s:conversion-gen=k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission
+// +k8s:defaulter-gen=TypeMeta
+
+// Package v1alpha1 is the v1alpha1 version of the API.
+// +groupName=webhook.admission.apiserver.k8s.io
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1alpha1/doc.go]: Please update this line to `apiserver.config.k8s.io`.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/kubeconfig.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/kubeconfig.go]: Then can we just remove `registry` and `groupFactoryRegistry`? How about
```golang
var (
scheme  = runtime.NewScheme()
codecs = serializer.NewCodecFactory(scheme)
)

func Init() {
  webhookadmission.AddToScheme(scheme)
  webhookadmissionv1alpha1.AddToScheme(scheme)
}
```

File: staging/src/k8s.io/apiserver/pkg/admission/config.go [hunk 9]
```diff
@@ -119,25 +110,29 @@ func ReadAdmissionConfiguration(pluginNames []string, configFilePath string) (Co
...
 }
 
 type configProvider struct {
 	config *apiserver.AdmissionConfiguration
+	scheme *runtime.Scheme
 }
 
 // GetAdmissionPluginConfigurationFor returns a reader that holds the admission plugin configuration.
-func GetAdmissionPluginConfigurationFor(pluginCfg apiserver.AdmissionPluginConfiguration) (io.Reader, error) {
+func GetAdmissionPluginConfigurationFor(pluginCfg apiserver.AdmissionPluginConfiguration, scheme *runtime.Scheme) (io.Reader, error) {
```
[the remark this answers is not in the document]
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/admission/config.go]: > Can this function be private?

Exposed if someone wants to write their own resolver.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/types.go (hunk 5 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/types.go]: I think it depends on how the cluster is configured. Shall the path be absolute?

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/types.go]: for now, documenting as absolute seems ok. post-1.9, I'd like to see sufficient information about the source of the config made part of the admission signature so admission plugins can resolve file references relative to the config file.

**Pull request merged**